### PR TITLE
[JENKINS-58494] - Remove dependency on Jenkins CLI from the core

### PR DESCRIPTION
**Pull request by oleg-nenashev**
From what I see, such change is required to fix https://issues.jenkins-ci.org/browse/JENKINS-58483 . See [JENKINS-58494](https://issues.jenkins-ci.org/browse/JENKINS-58494) for a proposal. Jenkins core currently includes Jenkins CLI as a dependency, and hence there is no way to split the CLI code and the Core codebase. There is little justification to keep the dependency, and IMHO we should not do so.

With this change, jenkins.py will have just another regression due to the removal of the deprecated `CliTransportAuthenticator` class. But I doubt this code is alive after JEP-200 anyway

What was changed here?

- [x] The dependency was removed from the core
- [x] CLI jar discovery was reworked to avoid dependencies on the class
- [x] Deprecated `CliTransportAuthenticator` was removed
- [x] QuotedStringTokenizer was copy-pasted and added to the core to avoid massive compatibility changes in the downstream plugins
- [ ] Plain CLI protocol fixes

I cannot say for sure that anything except jenkins.py does not depend on the CLI classes. Needs careful reviews

### Proposed changelog entries

* RFE: Remove the dependency on Jenkins CLI from the Jenkins Core

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
Not gonna work until `PlainCLIProtocol` is fixed. There are 2 ways:

* Copy-paste `PlainCLIProtocol#EitherSide` so that master and agent have their own implementations
* Introduce a new library which would hold only the relevant classes to be shared between CLI and the core
* Fallback to the original code and just do the dependency exclusions

File: core/src/main/java/hudson/cli/CliTransportAuthenticator.java [hunk 1]
```diff
@@ -1,40 +0,0 @@
...
-
-import hudson.ExtensionList;
-import hudson.ExtensionPoint;
-import hudson.remoting.Channel;
-
-/**
- * @deprecated No longer used.
- */
-@Deprecated
-public abstract class CliTransportAuthenticator implements ExtensionPoint {
```
Comment (jglick) [inline core/src/main/java/hudson/cli/CliTransportAuthenticator.java]: Fine, this was removed in `ssh-cli-auth` 1.5, which we now bundle.

File: cli/src/main/java/hudson/cli/CLIQuotedStringTokenizer.java [hunk 2]
```diff
@@ -0,0 +1,562 @@
...
+ * the behaviour is the same, except that single and double quoted
+ * string values are recognized.
+ * Delimiters within quotes are not considered delimiters.
+ * Quotes can be escaped with '\'.
+ *
+ * @see java.util.StringTokenizer
+ * @author Greg Wilkins (gregw)
+ */
+public class CLIQuotedStringTokenizer
+    extends StringTokenizer
```
Comment (jglick) [inline cli/src/main/java/hudson/cli/CLIQuotedStringTokenizer.java]: Make a note of where in Jenkins sources it was copied from.

Did you check for some Commons utility to do something sufficiently similar?

File: core/src/main/java/hudson/cli/DiagnosedStreamCorruptionException.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
+package hudson.cli;
+
+import java.io.PrintWriter;
+import java.io.StreamCorruptedException;
+import java.io.StringWriter;
+
+// TODO COPIED FROM hudson.remoting
```
Comment (jglick) [inline core/src/main/java/hudson/cli/DiagnosedStreamCorruptionException.java]: So why are we copying it? If it is just not `public` in `remoting`, then file a PR to fix that please…

File: core/src/main/java/hudson/cli/FlightRecorderInputStream.java [hunk 4]
```diff
@@ -0,0 +1,191 @@
+package hudson.cli;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Arrays;
+
+// TODO COPIED FROM hudson.remoting
```
Comment (jglick) [inline core/src/main/java/hudson/cli/FlightRecorderInputStream.java]: ditto

File: core/src/main/java/hudson/cli/HexDump.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
+package hudson.cli;
+
+// TODO COPIED FROM hudson.remoting
```
Comment (jglick) [inline core/src/main/java/hudson/cli/HexDump.java]: ditto

File: core/src/main/java/hudson/model/Slave.java [hunk 6]
```diff
@@ -407,19 +406,17 @@ public URL getURL() throws IOException {
...
-                File cliJar = Which.jarFile(CLI.class);
-                if (cliJar.isFile()) {
-                    name = "jenkins-cli.jar";
-                } else {
-                    URL res = findExecutableJar(cliJar, CLI.class);
-                    if (res != null) {
-                        return res;
-                    }
+                name = "jenkins-cli.jar";
+                // We know that Remoting CLI and Jenkins CLI are located in the same directory.
```
Comment (jglick) [inline core/src/main/java/hudson/model/Slave.java]: Not sure I am following where you are making use of that assumption.

File: core/src/main/java/hudson/util/QuotedStringTokenizer.java [hunk 7]
```diff
@@ -49,7 +50,10 @@
  * Delimiters within quotes are not considered delimiters.
  * Quotes can be escaped with '\'.
  *
- * @see java.util.StringTokenizer
+ * This class was originally in the CLI package, but it was moved to the Jenkins Core as a part of
```
Comment (jglick) [inline core/src/main/java/hudson/util/QuotedStringTokenizer.java]: As part of…?

Why is this comment important to a reader? Just omit it.

File: test/src/test/java/hudson/model/SlaveTest.java [hunk 8]
```diff
@@ -130,9 +130,7 @@ public void shouldNotEscapeJnlpSlavesResources() throws Exception {
         // Spot-check correct requests
         assertJnlpJarUrlIsAllowed(slave, "agent.jar");
         assertJnlpJarUrlIsAllowed(slave, "slave.jar");
-        assertJnlpJarUrlIsAllowed(slave, "remoting.jar");
-        assertJnlpJarUrlIsAllowed(slave, "jenkins-cli.jar");
-        assertJnlpJarUrlIsAllowed(slave, "hudson-cli.jar");
+       assertJnlpJarUrlIsAllowed(slave, "hudson-cli.jar");
```
Comment (jglick) [inline test/src/test/java/hudson/model/SlaveTest.java]: Why the reindentation?

File: test/src/test/java/hudson/model/SlaveTest.java [hunk 9]
```diff
@@ -152,6 +150,15 @@ public void shouldNotEscapeJnlpSlavesResources() throws Exception {
         assertJnlpJarUrlFails(slave, "./../foo/bar");
     }
 
+    @Issue("JENKINS-58494")
+    public void shouldResolveCLIJARsAsJNLPJars() throws Exception {
+        Slave slave = j.createSlave();
+
+        // Spot-check correct requests
```
Comment (jglick) [inline test/src/test/java/hudson/model/SlaveTest.java]: Why is this test being split in half?

File: core/src/main/java/hudson/cli/PlainCLIProtocol.java [hunk 10]
```diff
@@ -0,0 +1,347 @@
...
+ * Each side consists of frames starting with an {@code int} length,
+ * then a {@code byte} opcode, then any opcode-specific data.
+ * The length does not count the length field itself nor the opcode, so it is nonnegative.
+ */
+class PlainCLIProtocol {
+
+    static final Logger LOGGER = Logger.getLogger(PlainCLIProtocol.class.getName());
+
+    /** One-byte operation to send to the other side. */
+    private enum Op {
```
Comment (jglick) [inline core/src/main/java/hudson/cli/PlainCLIProtocol.java]: Not comfortable with the protocol definition and `EitherSide` being duplicated. Just introduce a new Maven module with `jar` packaging for the common code between `cli` and `jenkins-core`.

**Review by jglick — CHANGES_REQUESTED** [review]
Duplication of `PlainCLIProtocol` seems too dangerous.

**Comment by jglick** [comment]
> From what I see, such change is required to fix https://issues.jenkins-ci.org/browse/JENKINS-58483 .

#4111 claims to do so, so this PR description is not actually true any more, right?

**Comment by jglick** [comment]
Also, this does not compile in `core` (`HelpCommand`), and I suppose you will have to adjust POMs to make sure the CLI überjar is still included in the WAR, and the CLI JAR is still a `test`-scoped dependency of the `test` module.

**Comment by oleg-nenashev** [comment]
> #4111 claims to do so, so this PR description is not actually true any more, right?

This PR is a more aggressive version of #4111 which removes the CLI dependnecy entirely. As you may see, it is not going to work as is, so #4111 is a priority for me ATM. Once that PR is merged, I will consider finalizing this one

**Comment by oleg-nenashev** [comment]
@jenkinsci/core Would it make sense to have a new `jenkins-cli-base-lib` (or whatever) which stores the code shared between the Master-side and CLI side? It might be a submodule here or a new library repository

**Comment by fcojfernandez** [comment]
My vote is for an external library with the share code there. Instead of having a dependency on `cli` both parts can have a dependency on that library

**Comment by jglick** [comment]
> Would it make sense to have a new `jenkins-cli-base-lib` (or whatever) which stores the code shared between the Master-side and CLI side?

That was my recommendation.

> It might be a submodule here or a new library repository

Submodule. There is no reason I can see to move this code to a different repository, unless you plan to move the whole CLI client as well.

**Review by ndeloof — APPROVED** [review]

File: cli/src/main/java/hudson/cli/CLIQuotedStringTokenizer.java [hunk 11]
```diff
@@ -0,0 +1,562 @@
...
+ * This class is a copy of the java.util.StringTokenizer API and
+ * the behaviour is the same, except that single and double quoted
+ * string values are recognized.
+ * Delimiters within quotes are not considered delimiters.
+ * Quotes can be escaped with '\'.
+ *
+ * @see java.util.StringTokenizer
+ * @author Greg Wilkins (gregw)
+ */
+public class CLIQuotedStringTokenizer
```
Comment (daniel-beck) [inline cli/src/main/java/hudson/cli/CLIQuotedStringTokenizer.java]: `@Restricted`?

**Comment by varyvol** [comment]
@oleg-nenashev  do you intend to keep working on this? Otherwise we could propose it for close and it can be reopen when necessary.

**Comment by oleg-nenashev** [comment]
No plans to work on it in the foreseeable future. The implementation will need a full rework anyway, hence closing

**Pull request closed without merging**